src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java

```diff
@@ -0,0 +1,170 @@
+        }
+        final List<Allele> allelesToEmit = vc.getAlleles();
+        if (allelesToEmit.size() <= 1) {
+            return;
+        }
+        final double[] prior = composePriorPseudoCounts(allelesToEmit.size());
+        final LikelihoodMatrix<GATKRead, Allele> sampleMatrix = likelihoods.sampleMatrix(likelihoods.indexOfSample(g.getSampleName()));
+        final LikelihoodMatrix<GATKRead, Allele> sampleMatrixForAlleles =
+                allelesToEmit.size() == likelihoods.numberOfAlleles()
+                        ? sampleMatrix : new SubsettedLikelihoodMatrix<>(sampleMatrix, allelesToEmit);
+
+        final double[] posteriors;
+        if (sampleMatrix.evidence().isEmpty()) {
+            posteriors = prior;
+        } else {
+            final RealMatrix lkMatrix = composeInputLikelihoodMatrix(likelihoods, sampleMatrixForAlleles);
+            final double[] weights = calculateWeights(lkMatrix);
+            posteriors = SomaticLikelihoodsEngine.alleleFractionsPosterior(lkMatrix, prior, weights);
+        }
+
+        final double[] frequencies = MathUtils.normalizeSumToOne(posteriors);
+        if (!keepPriorInCount) {
+            for (int i = 0; i < posteriors.length; i++) {
+                posteriors[i] -= prior[i];
+            }
+        }
+        gb.attribute(DEPTH_KEY, Arrays.stream(posteriors)
+                .mapToObj(DEPTH_FORMAT::format)
+                .collect(Collectors.joining(VCFConstants.INFO_FIELD_ARRAY_SEPARATOR)));
+        gb.attribute(FRACTION_KEY,Arrays.stream(frequencies)
+                .mapToObj(FRACTION_FORMAT::format)
+                .collect(Collectors.joining(VCFConstants.INFO_FIELD_ARRAY_SEPARATOR)));
+    }
+
+    private RealMatrix composeInputLikelihoodMatrix(AlleleLikelihoods<GATKRead, Allele> likelihoods, LikelihoodMatrix<GATKRead, Allele> sampleMatrixForAlleles) {
+        final RealMatrix lkMatrix;
+        sampleMatrixForAlleles.asRealMatrix().copy();
+        if (!likelihoods.isNaturalLog()) {
+            lkMatrix = sampleMatrixForAlleles.asRealMatrix().copy();
+            final RealMatrixChangingVisitor log10ToLnTransformer = new DefaultRealMatrixChangingVisitor() {
+                public double visit(int row, int column, double value) {
+                    return value * MathUtils.LOG_10;
+                }
+            };
+            lkMatrix.walkInOptimizedOrder(log10ToLnTransformer);
+        } else {
+            lkMatrix = sampleMatrixForAlleles.asRealMatrix();
+        }
+        return lkMatrix;
+    }
+
+    private double[] calculateWeights(RealMatrix lkMatrix) {
+        if (weightDecay <= 0) {
+            return null;
+        } else {
+            final double[] result = new double[lkMatrix.getColumnDimension()];
+            for (int i = 0; i < result.length; i++) {
+                double best = lkMatrix.getEntry(0, i);
+                double secondBest = Double.NEGATIVE_INFINITY;
+                for (int j = 1; j < lkMatrix.getRowDimension(); j++) {
```

Comment:
This makes me really nervous.  Why, does i start with 0, and j with 1?

Reply:
i is standard 
j if you look at the other two previous lines in the outer for loop (```double best = ...``` and ```double secondBest = ```) the work for j == 0 is done, that is why we jump straight to j = 1

Reply:
I see, ty.